src/test/java/io/vertx/core/http/HttpBandwidthLimitingTest.java

```diff
@@ -0,0 +1,289 @@
+    HttpServer testServer = serverFactory.apply(vertx);
+    testServer.requestHandler(HANDLERS.getFile(sampleF));
+    startServer(testServer);
+
+    watch.start();
+    HttpClient testClient = clientFactory.apply(vertx);
+    testClient.request(HttpMethod.GET, testServer.actualPort(), DEFAULT_HTTP_HOST,"/get-file",
+                       req -> req.result().send(resp -> {
+                         resp.result().bodyHandler(r -> testComplete());
+                       }));
+    await();
+    watch.stop();
+
+    assertTrue(watch.getTime() > 1000);
+  }
+
+  @Test
+  public void dataUploadIsThrottled() throws Exception {
+    Buffer expectedBuffer = TestUtils.randomBuffer((TEST_CONTENT_SIZE));
+
+    StopWatch watch = new StopWatch();
+    HttpServer testServer = serverFactory.apply(vertx);
+    testServer.requestHandler(HANDLERS.bufferWrite(expectedBuffer));
+    startServer(testServer);
+
+    watch.start();
+    HttpClient testClient = clientFactory.apply(vertx);
+    write(expectedBuffer, testServer, testClient);
+    await();
+    watch.stop();
+
+    assertTrue(watch.getTime() > expectedTimeMillis(TEST_CONTENT_SIZE, INBOUND_LIMIT));
+  }
+
+  @Test
+  public void fileUploadIsThrottled() throws Exception {
+    StopWatch watch = new StopWatch();
+    HttpServer testServer = serverFactory.apply(vertx);
+    testServer.requestHandler(HANDLERS.uploadFile(sampleF));
+    startServer(testServer);
+
+    watch.start();
+    HttpClient testClient = clientFactory.apply(vertx);
+    upload(testServer, testClient, sampleF);
+    await();
+    watch.stop();
+
+    assertTrue( watch.getTime() > 1000);
+  }
+
+  @Test
+  public void testSendFileTrafficShapedWithSharedServers() throws InterruptedException {
+    int numEventLoops = 4; // We start a shared TCP server with 4 event-loops
+    Future<String> listenLatch = vertx.deployVerticle(() -> new AbstractVerticle() {
+      @Override
+      public void start(Promise<Void> startPromise) {
+        HttpServer testServer = serverFactory.apply(vertx);
+        testServer.requestHandler(HANDLERS.getFile(sampleF))
+                  .listen(DEFAULT_HTTP_PORT, DEFAULT_HTTP_HOST).<Void>mapEmpty().onComplete(startPromise);
+      }
+    }, new DeploymentOptions().setInstances(numEventLoops));
+
+    HttpClient testClient = clientFactory.apply(vertx);
+    StopWatch watch = new StopWatch();
+    CountDownLatch waitForResponse = new CountDownLatch(4);
+    listenLatch.onComplete(v -> {
+      watch.start();
+      for (int i=0; i<4; i++) {
+        testClient.request(HttpMethod.GET, DEFAULT_HTTP_PORT, DEFAULT_HTTP_HOST,"/get-file", req -> {
+          HttpClientRequest result = req.result();
+          result.send(resp -> {
+            resp.result().bodyHandler(r -> waitForResponse.countDown());
+          });
+        });
+      }
+    });
+    awaitLatch(waitForResponse);
+    watch.stop();
+    assertTrue(watch.getTime() > 6000); // because there are simultaneous 4 requests
```

Comment:
we need here to use `Assert.assertTrue(...)` instead of `assert(...)` because `await()` was called before and the async test supposes the test is finished and no error can be reported